memory: extract rangeFilterExpr into expression package

paralin opened the pull request:
rangeFilterExpr contains a complex set of logic to build a sql expression given the list of sql ranges and the list of expressions on the index.

Extract the majority of this function into NewRangeFilterExpr in the expression package. Replace the Or and And helper functions with JoinOr and JoinAnd.

Update the call in the memory/ package to call the new expression function.

Fix JoinOr and JoinAnd to check if the expressions are nil.

Remove ineffective nil check in JoinAnd.

Move nil checks into NewOr and NewAnd.

Add comment explaining parameters to NewFilterRangeExpr.

paralin commented:
Includes changes requested in #1606 

Thanks for the review @fulghum

fulghum commented on sql/expression/logic.go:
```diff
@@ -137,9 +143,31 @@ var _ sql.CollationCoercible = (*Or)(nil)
 	return &Or{BinaryExpression{Left: left, Right: right}}
 }
 
+// JoinOr joins several expressions with Or.
+func JoinOr(exprs ...sql.Expression) sql.Expression {
```
Could you please add a unit test for this new exported function? I'm happy to accept this change, but we need test coverage. It should be trivial to add to the existing unit tests in `logic_test.go`.

paralin replied:
Added!

fulghum reviewed (commented):
Hey @paralin, thanks for opening this; apologies for the churn earlier this week with closing the old PR. Code changes look good, but something in here is causing CI tests to segfault:
https://github.com/dolthub/go-mysql-server/actions/runs/4738411832/jobs/8412238529?pr=1720

This should be good to go once:
- all CI tests are passing cleanly
- there's unit test coverage for the new `JoinOr` function

paralin commented:
@fulghum The nil expressions are causing the segfault, apparently it's not OK to have a nil expression in a `&And{BinaryExpression{Left: left, Right: right}}`

This is why the checking was there before to skip nil expressions, it seems.

However the code before was not sufficient either because it would not handle the case where all the expressions were nil.

I guess I have a broader question which is - why are there nil expressions floating around anyway, and what is the output of And(nil, nil) supposed to be? What's the output of And(nil, expr)? (Seems like it should just be expr). Or what is the output of Or(nil, nil) or Or(nil, expr)? And why are these nils appearing at all?

Would love if someone with more experience with this could answer.

fulghum commented:
Thanks for digging in @paralin. I am a bit surprised that the `nil` expressions are being passed around. I won't have time to dig into this today, but can try to help dig a little deeper tomorrow. @max-hoffman is our lead for the analyzer, and may have a moment to chime in.

paralin commented:
@max-hoffman That was how `joinOr` and `joinAnd` were written before, it was changed due to a PR review comment.

Added the nil checks back and added an extra test call to JoinOr and JoinAnd.

Seems the tests pass now!

fulghum reviewed (approved):
Thanks for all the work on this one @paralin! 🙏  Merging it in now for ya...

paralin commented:
@fulghum great! thanks!
